### PR TITLE
Add Nim language support

### PR DESCRIPTION
**Pull request by omentic**
Supersedes #3320.

### CODE REVIEW

**Comment by sigmaSd** [comment]
The indents queries are not correct I think, nvim doesn't have extend or outdent, but it have indent and dedent (and maybe more)

**Comment by sigmaSd** [comment]
Also you maybe want to consider adding textobjects.scm, user can use it with https://github.com/nvim-treesitter/nvim-treesitter-textobjects (it will need some changes to the capture names though)

**Comment by amaanq** [comment]
This is not how nvim-treesitter writes queries, please refer to https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights for acceptable capture groups

**Comment by clason** [comment]
> Also you maybe want to consider adding textobjects.scm, user can use it with https://github.com/nvim-treesitter/nvim-treesitter-textobjects (it will need some changes to the capture names though)

These are not tracked in this repo; that is a separate module.

**Comment by theHamsta** [comment]
The author of tree-sitter-nim writes in his repo (and he told when I was reporting the SEGFAULT when my PR was created) that he doesn't want his software to be deployed even though that is his final goal once he considers it to be ready. It would prefer to respect that desire if that's still his attitude.

@aMOPel

**Comment by aMOPel** [comment]
> that he doesn't want his software to be deployed even though that is his final goal once he considers it to be ready

Oh, I'm sorry. Then we had a misunderstanding. I just assumed the parser isn't useful unless it is complete. 

Feel free to use it.

I am not planning to complete it currently however. I got stuck on the last few syntax nodes and I am missing some crucial understanding for some aspects of tree sitter I believe.

https://github.com/alaviss/tree-sitter-nim/tree/rewrite

Is currently taking another shot at it.

**Comment by amaanq** [comment]
> I am not planning to complete it currently however. I got stuck on the last few syntax nodes and I am missing some crucial understanding for some aspects of tree sitter I believe.

I'll put this as a TODO of mine sometime later..would be cool to have.

**Comment by j-james** [comment]
Thanks @amaanq, I wrote this originally for Helix and forgot they used different capture groups. These should work with Neovim now.

I haven't yet tested this in Neovim itself: how do you recommend doing so?

**Comment by amaanq** [comment]
> Thanks @amaanq, I wrote this originally for Helix and forgot they used different capture groups. These should work with Neovim now.
> 
> I haven't yet tested this in Neovim itself: how do you recommend doing so?

Install the plugin, in setup add something like
```lua
local parser_configs = require("nvim-treesitter.parsers").get_parser_configs()
parser_configs.nim = {
  install_info = {
    url = "~/projects/treesitter/tree-sitter-luap",
    files = { "src/parser.c" }, -- and "src/scanner.c(c)" if applicable
  },
}
```
and copy the queries to the nvim-treesitter install location at queries/nim/

**Comment by j-james** [comment]
@amaanq If you're going to look at the parser, you should probably look at alaviss's rewrite over at https://github.com/alaviss/tree-sitter-nim/tree/rewrite : it's less complete but structured to reflect the language better. When it's complete I was planning on rewriting the queries for it regardless.

**Comment by j-james** [comment]
After some more testing I'm a little apprehensive about merging this. tree-sitter-nim errors out sometimes, which leaves chunks of code unhighlighted. I think it may be better to merge https://github.com/alaviss/nim.nvim/blob/master/syntax/nim.vim into neovim itself and wait on semantic highlighting until the rewrite is complete.

**Comment by amaanq** [comment]
> After some more testing I'm a little apprehensive about merging this. tree-sitter-nim errors out sometimes, which leaves chunks of code unhighlighted.

My initial thoughts were this, this parser is nowhere near suitable for usage rn imo, it is not complete or near complete. 

>  I think it may be better to merge https://github.com/alaviss/nim.nvim/blob/master/syntax/nim.vim into neovim itself and wait on semantic highlighting until the rewrite is complete.

Submit it to vim as a syntax file, then backport it as a vim patch to neovim if you'd like. I'll take a stab at this when I get time, working on Hare rn..

**Comment by amaanq** [comment]
Issues I see with it, let alone it being incomplete imo (though these can be overlooked, it just provides a worse experience as it can be laggier)

#define STATE_COUNT 4583
#define LARGE_STATE_COUNT 2895 
Causes it to currently compile at ~2.5mb for me

Author doesn't seem to understand how the external scanner works, as it is practically the python external scanner copied with bits commented out (perhaps being the cause of the segfault before)

The parser takes a while to scan the core repo, and 1364 files fail (though perhaps a lot are just because of those "complex type declarations")

The grammar has quite some conflicts, and I fear that while certain bits of code *may* parse ok, a lot will parse **incorrectly** (think a block parsing as a struct due to ambiguity, and vice versa)

So, at best this should be 100% marked experimental, at worse I'll rewrite it myself in a bit.

**Comment by aMOPel** [comment]
> Author doesn't seem to understand how the external scanner works, as it is practically the python external scanner copied with bits commented out (perhaps being the cause of the segfault before)

I did copy the python external scanner as a starting point but modified it where necessary. Admittedly I don't understand everything on a deeper level, but I added functionality for multiline strings, and string interpolations (`&"{x}"`), which both work reasonably well.

The segfault was indeed caused by a bug I introduced in the external scanner. I managed to fix it, though.

> The parser takes a while to scan the core repo, and 1364 files fail (though perhaps a lot are just because of those "complex type declarations")

I fixed these type declaration issues yesterday. There were a few bugs in there but now it seems to be doing fine with type declarations. I added a bunch of test cases for those bugs as well.

The parser is still not complete, as it's missing support for complex expressions (like if-expressions) and support for arbitrary parentheses (like `- (5+5)`).

> #define STATE_COUNT 4583
> #define LARGE_STATE_COUNT 2895
> Causes it to currently compile at ~2.5mb for me

I am fully on board with the parser being enormous and thus slower. The grammar structure is close to the official grammar spec, which helped me writing, but made for a bloated product.

**Comment by j-james** [comment]
> I fixed these type declaration issues yesterday.

Woah, that's way better! I'll do some more testing and send any issues your way but I think you might have fixed everything?

**Comment by sigmaSd** [comment]
seems to work well

Only thing I notice is that the brackets sometimes only one of the pair gets highlighted  (but there is no error in the syntax tree)
image

**Comment by j-james** [comment]
Thanks for the report. The other bugs I know of are:
- entries in `std/[one, two]` get highlighted as variables instead of namespaces
- the variables a and b in `seq[tuple[a, b: int]]` are accidentally matched as types
- indents don't work correctly I think?

**Comment by aMOPel** [comment]
I rewrote the highlight queries to be compliant with the nvim treesitter expectations.

https://github.com/aMOPel/tree-sitter-nim/blob/nvim_treesitter_queries/queries/nvim/highlights.scm

Gonna polish it a little more and maybe improve indents as well before pushing it here.
Could also add regex injection.

I wrote some locals queries, too, but I didn't really know how to test them yet. Maybe @amaanq or somebody else could help me out there.

**Comment by tobealive** [comment]
Thanks a ton for that work! I'm using it already :+1: 

What a pleasure to finally have treesitter highlighting. The only tiny issue I'm noticing so far is that it doesn't work for block comments.

**Comment by aMOPel** [comment]
> The only tiny issue I'm noticing so far is that it doesn't work for block comments.

There are plenty of issues. Many unhandled corner cases. Also

> The parser is still not complete, as it's missing support for complex expressions (like if-expressions) and support for arbitrary parentheses (like `- (5+5)`).

But all in all it does about 95% of Nim's syntax features and works good enough for most cases.
To make it perfect, however, a rewrite is inevitable I believe.

**Comment by theHamsta** [comment]
Should we merge this already to have something to improve upon?

**Comment by aMOPel** [comment]
> The only tiny issue I'm noticing so far is that it doesn't work for block comments.

Just checked and I did indeed forget to add the multiline comments to the queries. Thanks!

**Comment by tobealive** [comment]
> The comment I sent was mainly focused on highlighting. 

Thanks to you! @aMOPel 

---

With the merge consideration I want to add a detail.
My last comment, that it's already in use might have sounded like there are no issues. But I'm actually a bad role model in the way I'm using it.

To avoid query errors related to indents, I excluded indent queries in nvim-treesitter/indent.lua for Nim. - This might be a common way of dealing with errors in Microsoft products, but it probably should not be used here. 

So my first comment might have been a bit misleading, but it was mainly about the highlighting, which is definitely at a level that can be merged and improved upon. The rest is almost as mature for a first merge as it looks.

**Comment by aMOPel** [comment]
I haven't looked in-depth at the indent.scm yet. I personally didn't use tree sitter based indent so far, so it didn't have a high priority for me. Text objects would be way me interesting to me personally.

**Comment by theHamsta** [comment]
> @theHamsta are you talking about my rewrite? How would I get it from my parser repo to this PR repo? Do I have to make a PR to https://github.com/j-james/nvim-treesitter or is there another way?

No, just this PR. I forgot that this PR used a fork. Did you agree on which repository should be the upstream? Probably yours @aMOPel

**Comment by aMOPel** [comment]
> Did you agree on which repository should be the upstream? Probably yours @aMOPel

Didn't hear from @j-james in a few days now.

**Comment by aMOPel** [comment]
> I can pull them into this fork.

making a pr to your fork rn

**Comment by j-james** [comment]
Sounds good, I'm also going to try and fix up the indent queries.

**Comment by j-james** [comment]
Hmm, weird, my queries testing setup (not nvim) works for highlights.scm but doesn't work for indent.scm? Strange.

**Comment by amaanq** [comment]
When updating indents.scm locally you have to reopen the editor (at least in my experience) for new indent logic to work.

**Comment by j-james** [comment]
I've been testing in Helix since the queries were originally written for their tree-sitter / LSP integration so I haven't gotten a Neovim testing environment set up. Maybe I should do that... but for now this can be merged as-is and I'll PR the indent queries separately.

**Comment by j-james** [comment]
@theHamsta mind re-triggering the CI and merging this?

File: queries/nim/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+(primary
+  (primarySuffix
+    (qualifiedSuffix
+      (symbol (ident) @function.call)))
+  . (primarySuffix (cmdCall)))
+; must come after @field
+
+(primary
+  (primarySuffix (qualifiedSuffix (symbol) @function.call))
+  . (primarySuffix (functionCall ["(" ")"] @function.call)))
```
Comment (theHamsta) [inline queries/nim/highlights.scm:247]: Same here?

**Review by theHamsta — COMMENTED** [review]
Apparent from that detail, this looks good. The indents might have soon a refactor to get a `@indent.`

File: queries/nim/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+] @float
+; floating-point number literals
+
+(routine (symbol [(ident) (operator)] @function))
+(routine (paramList ["(" ")"] @function))
+(routine "=" @function)
+; function definitions
+
+; primaryPrefix breaks function.call query
+(castExpr ["(" ")"] @function.call)
```
[the remark this answers is not in the document]
Comment (j-james) [reply queries/nim/highlights.scm:108]: cc. @aMOPel

[the remark this answers is not in the document]
Comment (aMOPel) [reply queries/nim/highlights.scm:108]: Thought it looked good. It's being done for all routine definitions and calls and includes the `( )` and `=` for definitions.

Also the `( )` of object constructors is captured.

But if you say it's not the tree-sitter way, we can drop it.

[the remark this answers is not in the document]
Comment (j-james) [reply queries/nim/highlights.scm:108]: @theHamsta Is this fine to merge or should we remove it?

[the remark this answers is not in the document]
Comment (aMOPel) [reply queries/nim/highlights.scm:108]: @j-james I already removed it made some additional changes on my main branch. You can pull it in if you want. (I adapted my highlights.scm to you formatting, so diffing should be easier)

[the remark this answers is not in the document]
Comment (theHamsta) [reply queries/nim/highlights.scm:108]: Should be removed to be consistent with other languages. Semantically colored braces could be a cool optional feature but probably nothing for a default experience

**Comment by hackeryarn** [comment]
@omentic do you need any help or anything to get this over the finish line? I would be more than happy to help in any way possible. This would be really great to get merged in. All the current syntax highlighters are broken in various way, after the new Nim release, since they depend on an outdated `nimsuggest` implementation.

**Comment by aMOPel** [comment]
@hackeryarn 
This PR has fallen asleep, due to various reasons. The most important being, that my grammar doesn't do a good enough job in the end. This PR would have only been a temporary solution  anyway.

In the mean time, https://github.com/alaviss/tree-sitter-nim has made some great progress and is supposedly able to parse [97.77%](https://github.com/alaviss/tree-sitter-nim/releases/tag/0.0.3) of the compiler code.

Apparently some things aren't quite perfect yet and the parser is still quite large, but overall that repo is the way forward.

If you want to help, it would be best directed towards that repo, probably doing more testing, giving feedback and/or writing queries.

**Comment by hackeryarn** [comment]
@aMOPel Thank you for the insight! I will take a look at that repo and see where I can help.

**Comment by omentic** [comment]
Superceded by #5556

**Pull request closed without merging**